### PR TITLE
WIP: fix for null oopMap

### PR DESCRIPTION
**Pull request by offamitkumar**
Doesn't fix the issue as of now, but you may want to pick it from here...

### CODE REVIEW

**Comment by offamitkumar** [comment]
`wget/curl` diff file from here: https://patch-diff.githubusercontent.com/raw/offamitkumar/jdk/pull/3.diff

**Comment by offamitkumar** [comment]
@backwaterred could you make it to merge by today, with our personal master branch ???

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 1]
```diff
@@ -1592,23 +1592,23 @@ static void gen_continuation_enter(MacroAssembler* masm,
...
 
+// Kills Z_R1, Z_R0
 static void gen_continuation_yield(MacroAssembler* masm,
                                    const VMRegPair* regs,
                                    OopMapSet* oop_maps,
                                    int& frame_complete,
                                    int& framesize_words,
                                    int& compiled_entry_offset) {
   Register Rtmp = Z_R0_scratch;
+  const int framesize_bytes = frame::z_abi_160_size; /*I think pushing a frame is not a good idea*/
```
Comment (backwaterred) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: > /*I think pushing a frame is not a good idea*/

Why not a good idea? In my understanding we need one for the oopmap to work.

RR's suggestion was to use `const int framesize_bytes = __ push_frame_abi160(0);` to set the frame size. If you check the implementation of the push_frame procedure, it actually takes care of alignment as well, then returns the final size. I thought this was a nice improvement over the way ppc does it.

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 2]
```diff
@@ -1618,7 +1618,7 @@ static void gen_continuation_yield(MacroAssembler* masm,
   // We use this nop for fast CodeBlob lookup, associate the OopMap
   // with it right away.
   __ post_call_nop();
-  OopMap* map = new OopMap(framesize_bytes / VMRegImpl::stack_slot_size, 1);
+  OopMap* map = new OopMap(framesize_bytes / VMRegImpl::stack_slot_size, 0);
```
Comment (backwaterred) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp:1620]: Niiiice :-)

**Review by backwaterred — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp (hunk 1 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: I think we are already pushing the frame at Line 1611

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp [hunk 3]
```diff
@@ -488,12 +488,15 @@ FreezeBase::FreezeBase(JavaThread* thread, ContinuationWrapper& cont, intptr_t*
...
+#if defined(PPC64)
   static const int doYield_stub_frame_size = frame::native_abi_reg_args_size >> LogBytesPerWord;
+#elif defined(S390)
+  static const int doYield_stub_frame_size = frame::z_abi_160_base_size >> LogBytesPerWord;
+#else
+  static const int doYield_stub_frame_size = frame::metadata_words;
 #endif
-  assert(SharedRuntime::cont_doYield_stub()->frame_size() == doYield_stub_frame_size, "");
+  assert(SharedRuntime::cont_doYield_stub()->frame_size() == doYield_stub_frame_size,
+         "SharedRuntime::cont_doYield_stub()->frame_size(): %d and doYield_stub_frame_size: %d must be equal", SharedRuntime::cont_doYield_stub()->frame_size(), doYield_stub_frame_size);
```
Comment (offamitkumar) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:499]: I think we can improve this message, not sure how. Any better suggestion ?

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp (hunk 1 above)
Comment (backwaterred) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: Yeah, but if we need to push a frame (and are pushing it later). why would we push changes that suggest it's a bad idea?

Comment (backwaterred) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: Also, I'm a fan of keeping the changes that RR suggested here.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp (hunk 3 above)
Comment (backwaterred) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:499]: No suggestions. I think we may end up reverting common code changes to assert statements (or doing them in a separate pass).

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp (hunk 1 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: >Yeah, but if we need to push a frame (and are pushing it later). why would we push changes that suggest it's a bad idea? 

removed..

>I'm a fan of keeping the changes that RR suggested here.


which change I removed ?

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 4]
```diff
@@ -1592,23 +1592,23 @@ static void gen_continuation_enter(MacroAssembler* masm,
...
 
   address start = __ pc();
   compiled_entry_offset = __ pc() - start;
 
-  // fixME:: save return PC and push entry frame
-  // not sure what to do with __ enter(), Other places where this was
-  // used s390x seems doing nothing.
-  __ save_return_pc(Z_R14);
-  const int framesize_bytes = __ push_frame_abi160(0);
-  framesize_words = framesize_bytes / wordSize;
```
Comment (backwaterred) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: Keep this

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 5]
```diff
@@ -1592,23 +1592,23 @@ static void gen_continuation_enter(MacroAssembler* masm,
...
+  // save return pc and push entry frame
+  __ save_return_pc();
+  __ push_frame_abi160(framesize_bytes);
```
Comment (backwaterred) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: Drop this

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 6]
```diff
@@ -1592,23 +1592,23 @@ static void gen_continuation_enter(MacroAssembler* masm,
...
+// Kills Z_R1, Z_R0
 static void gen_continuation_yield(MacroAssembler* masm,
                                    const VMRegPair* regs,
                                    OopMapSet* oop_maps,
                                    int& frame_complete,
                                    int& framesize_words,
                                    int& compiled_entry_offset) {
   Register Rtmp = Z_R0_scratch;
+  const int framesize_bytes = frame::z_abi_160_size;
+  framesize_words = framesize_bytes / wordSize;
```
Comment (backwaterred) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: Drop this

**Review by backwaterred — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp (hunk 1 above)
Comment (backwaterred) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: > > I'm a fan of keeping the changes that RR suggested here.

> which change I removed ?

I marked them in another review.

**Comment by backwaterred** [comment]
I think these changes are mostly good. But, I want to keep the changes that I made in my last PR. I tagged them in a review above.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp [hunk 7]
```diff
@@ -488,12 +488,15 @@ FreezeBase::FreezeBase(JavaThread* thread, ContinuationWrapper& cont, intptr_t*
 
   assert(_cont.chunk_invariant(), "");
   assert(!Interpreter::contains(_cont.entryPC()), "");
-#if !defined(PPC64) || defined(ZERO)
-  static const int doYield_stub_frame_size = frame::metadata_words;
```
Comment (backwaterred) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:492]: Should metadata words be changed to `sizeof(z_abi_160_base) >> ...` and be used here?

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp [hunk 8]
```diff
@@ -488,12 +488,15 @@ FreezeBase::FreezeBase(JavaThread* thread, ContinuationWrapper& cont, intptr_t*
...
-#else
+#if defined(PPC64)
```
Comment (backwaterred) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:491]: I think this needs to be `#if defined(PPC64) && !defined(ZERO)` to match the previous logic.

**Review by backwaterred — APPROVED** [review]

**Pull request merged**